pkg/fatfs/fatfs_diskio/mtd/mtd_diskio.c

```diff
@@ -0,0 +1,228 @@
+ * @param[in]  sector  Start sector in LBA
+ * @param[in]  count   Number of sectors to read
+ *
+ * @return             RES_OK if no error occurred
+ * @return             RES_ERROR if data wasn't read completely
+ * @return             RES_PARERR if prdv is invalid or has no mtd-driver set
+ */
+DRESULT disk_read(BYTE pdrv, BYTE *buff, DWORD sector, UINT count)
+{
+    DEBUG("disk_read: %d, %lu, %d\n", pdrv, sector, count);
+    if ( (pdrv >= _VOLUMES) || (fatfs_mtd_devs[pdrv]->driver == NULL) ) {
+        return RES_PARERR;
+    }
+
+    int res = mtd_read(fatfs_mtd_devs[pdrv], buff,
+                       sector * fatfs_mtd_devs[pdrv]->page_size,
+                       count * fatfs_mtd_devs[pdrv]->page_size);
+
+    if (res >= 0) {
+        uint32_t r_sect = ((unsigned)res)/fatfs_mtd_devs[pdrv]->page_size;
+        return ( (r_sect == count) ? RES_OK : RES_ERROR);
+    }
+
+    return RES_ERROR;
+}
+
+/**
+ * @brief              writes sectors to disk
+ *
+ * @param[in]  pdrv    Physical drive nmuber to identify the drive
+ * @param[in]  buff    Data to be written
+ * @param[in]  sector  Start sector in LBA
+ * @param[in]  count   Number of sectors to write
+ *
+ * @return             RES_OK if no error occurred
+ * @return             RES_ERROR if data wasn't written completely
+ * @return             RES_PARERR if prdv is invalid or has no mtd-driver set
+ */
+DRESULT disk_write(BYTE pdrv, const BYTE *buff, DWORD sector, UINT count)
+{
+    DEBUG("disk_write: %d, %lu, %d\n", pdrv, sector, count);
+    if ( (pdrv >= _VOLUMES) || (fatfs_mtd_devs[pdrv]->driver == NULL) ) {
+        return RES_PARERR;
+    }
+
+    /* erase memory before writing to it */
+    int res = mtd_erase(fatfs_mtd_devs[pdrv],
+                        sector * fatfs_mtd_devs[pdrv]->page_size,
+                        count * fatfs_mtd_devs[pdrv]->page_size);
+
+    if (res != 0) {
+        return RES_ERROR; /* erase failed! */
+    }
+
+    res = mtd_write(fatfs_mtd_devs[pdrv], buff,
+                    sector * fatfs_mtd_devs[pdrv]->page_size,
+                    count * fatfs_mtd_devs[pdrv]->page_size);
+
+    if (res >= 0) {
+        uint32_t w_sect = ((unsigned)res)/fatfs_mtd_devs[pdrv]->page_size;
+        return ((w_sect == count) ? RES_OK : RES_ERROR);
+    }
+
+    return RES_ERROR;
+}
+
+/**
+ * @brief                  perform miscellaneous low-level control functions
+ *
+ * @param[in]      pdrv    Physical drive nmuber (0.._VOLUMES-1)
+ * @param[in out]  cmd     Control code
+ * @param[in]      sector  Buffer to send/receive control data
+ *
+ * @return                 RES_OK if no error occurred
+ * @return                 RES_PARERR if cmd is unknown or
+ *                         prdv is either invalid or has no mtd-driver set
+ */
+DRESULT disk_ioctl(BYTE pdrv, BYTE cmd, void *buff)
+{
+    if ( (pdrv >= _VOLUMES) || (fatfs_mtd_devs[pdrv]->driver == NULL) ) {
+        return RES_PARERR;
+    }
+
+    switch (cmd) {
+        #if (_FS_READONLY == 0)
```

Comment:
I think preprocessor statements should no be indented, here and in some other places